### PR TITLE
Clearer fringe evaluation indicator

### LINKED ISSUE #2748: Evaluation indicators too subtle?
The evaluation indicators in the fringe do not stand out very clearly and many people are unaware of this very useful feedback because they do not notice them.

I have found the evaluation indicators in the fringe particularly difficult to spot using either of the Spacemacs themes.  They are almost invisible on the spacemacs-light theme

Currently, `lightgreen` is used with light themes and `darkgreen` is used for dark themes.  Could these values be swapped?  That may make the indicators more obvious.

https://github.com/clojure-emacs/cider/blob/master/cider-overlays.el#L132

I would be happy to test out alternatives if there is a way to set the foreground colour, preferably setting that in the `init.el` or `.spacemacs` file than creating a fork of the CIDER package.

### PR DESCRIPTION
**Pull request by jcaw**
Closes #2748

# Overview

Issue #2748 notes that the evaluation indicator in the fringe is very subtle. This PR makes three changes to address this, split into 3 commits:

1. Extract the bitmap displayed in the fringe into a user-configurable `custom`.
2. Inherit the indicator's color from the current theme, rather than hard-coding it.
3. Change the default bitmap used to `filled-rectangle`.

This is just an initial stab at changing the indicators, to prove the concept. Happy to rewrite as necessary. I haven't cleared the checklist yet - wanted to hear what you guys thought before formally cleaning things up.

# Current Look

The indicator is difficult to see in certain themes. Colors are hard-coded, which can result in an inconsistent palette. The bitmap itself creates a semi-transparent effect. Some examples:

before-spacemacs-dark
`spacemacs-dark`

before-spacemacs-light
`spacemacs-light`

before-nord
`doom-nord`

# Different Bitmap

Swapping to a different bitmap makes things a little bolder. I had a look through all the built-in bitmaps, in my opinion `filled-rectangle` looks best.

symbol-only-spacemacs-dark
`spacemacs-dark`

symbol-only-spacemacs-light
`spacemacs-light`

symbol-only-nord
`doom-nord`

# Dynamic Colors

Rather than expecting each theme to explicitly override the fringe colors, it would be better if we could dynamically adjust the color to match each theme. `diff-attach` is a built-in face which is almost always green, and should match the palette of the theme.

Inheriting the foreground from this face works across a wide range of themes:

both-spacemacs-dark
`spacemacs-dark`

both-spacemacs-light
`spacemacs-light`

both-nord
`doom-nord`

both-zenburn
`zenburn`

both-leuven
`leuven`

both-doom-one
`doom-one`

both-monokai
`monokai`

both-solarized-dark
`solarized-dark`

both-solarized-light
`solarized-light`

Many themes also set the background for `diff-attach`. Inheriting just the foreground reduces visual clutter, and seems to work consistently. We can make the operation safer with `:distant-foreground`, to defer to a hard-coded green if the diff foreground is not sufficiently visible.

## Scaling

An issue with `filled-rectangle` is it doesn't scale vertically. It seems the proportions are locked and the fringe is pegged to a specific pixel width. The original bitmap doesn't have this issue. For example:

large-text

I'm not sure if defining a new bitmap would allow us to adjust this property. Another solution might also be to adjust the background color of the fringe directly, rather than display a bitmap.

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](../.github/CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [ ] All tests are passing (`make test`)
- [ ] All code passes the linter (`make lint`) which is based on [`elisp-lint`](https://github.com/gonewest818/elisp-lint) and includes
  - [byte-compilation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Byte-Compilation.html), [`checkdoc`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Tips.html), [check-declare](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declaring-Functions.html), packaging metadata, indentation, and trailing whitespace checks.
- [ ] You've updated the [changelog](../blob/master/CHANGELOG.md) (if adding/changing user-visible functionality)
- [ ] You've updated the [user manual](../blob/master/doc) (if adding/changing user-visible functionality)

Thanks!

*If you're just starting out to hack on CIDER you might find this [section of its
manual][1] extremely useful.*

[1]: https://docs.cider.mx/cider/contributing/hacking.html

### CODE REVIEW

**Comment by bbatsov** [comment]
Nicely done! A couple of small remarks:

* the CI is failing now
* I think we chose the current bitmap with the idea that it would look different from what some other modes use in the fringe from VC status. 
* It'd be nice if this was also documented in the manual - somewhere in basic usage would be best IMO.

File: cider-overlays.el [hunk 1]
```diff
@@ -29,6 +29,7 @@
 (require 'subr-x)
 (require 'cider-compat)
 (require 'cl-lib)
+(require 'diff-mode)
```
Comment (bbatsov) [inline cider-overlays.el]: Why did you add this?

File: cider-overlays.el [hunk 2]
```diff
@@ -130,14 +131,32 @@ This function also removes itself from `post-command-hook'."
...
+The default choices are all quite aesthetically pleasing, and have no role
+in the fringe by default. See `fringe-bitmaps' in the Emacs manual for a
+full list of available bitmaps."
+  :type '(choice (const empty-line)
+                 (const filled-rectangle)
+                 (const filled-square)
+                 (const hollow-square)
+                 (const vertical-bar))
+  :group 'cider
+  :package-version '(cider . "0.23.1"))
```
Comment (bbatsov) [inline cider-overlays.el]: Here it should say "0.24.0".

File: cider-overlays.el (hunk 1 above)
Comment (jcaw) [reply cider-overlays.el]: The `diff-added` face comes from this (built-in) package.

File: cider-overlays.el (hunk 2 above)
Comment (jcaw) [reply cider-overlays.el]: Perfect. Wasn't sure which version to use.

File: cider-overlays.el (hunk 1 above)
Comment (bbatsov) [reply cider-overlays.el]: Got it. Please add a comment next to the require, so it's clear.

**Comment by jcaw** [comment]
Might not be able to get back to this for a bit, but I will eventually.

**Comment by bbatsov** [comment]
No worries.

**Comment by jr0cket** [comment]
@jcaw thanks for the work on the PR for evaluation markers in the fringe for CIDER.  Are you able to complete this PR or would you like some assistance?

Thank you.

**Comment by jcaw** [comment]
Thanks for the reminder - I should be able to get back to it this week.

**Comment by bbatsov** [comment]
@jcaw The build is failing, because by Emacs conventions you should put two spaces after `.` in docstrings and comments.

**Comment by jr0cket** [comment]
@jcaw need any help with this PR?

**Comment by bbatsov** [comment]
@jr0cket seem you can step in here. :-)

**Comment by jcaw** [comment]
Sorry for the radio silence on this one - that's absolutely fine. Feel free to take over, and let me know if you have any questions.
